modified elasticsearch reporting module

swackhamer opened the pull request:
Couple of changes:
Implemented a dynamic elasticsearch template that matches on all strings and explicitly maps them to "not_analyzed".  This ensures that elasticsearch does not split on whitespace or other delimiters such as dashes or dots when performing searches or aggregations on that field.  More information here: https://www.elastic.co/guide/en/elasticsearch/guide/current/mapping-intro.html#custom-field-mappings

Added report_time which is the time in which the report gets indexed into elasticsearch.  This is very helpful for kibana and visualizing reports over time periods.

Added task_id to map the elasticsearch report back to the Cuckoo task_id

Reports now index to cuckoo-YYYY-mm-dd which allows partitioning of data.

Default shard count is now set to 1 primary and 1 replica shard.  This will prevent shard overallocation and will increase performance.  This setting should work even for the largest Cuckoo server implementations out there (limit of 1 shard is 4 billion documents).

razuz commented:
just a thought - why not add cuckoo hostname also to the document ? 
AFAIK currently it will not report which host made the analyze. If hostname is there then it would be possible to have one ES across cuckoo cluster and know which cuckoo instance holds the data

swackhamer commented:
I was thinking about that but did not know where the best place to pull that information is as there is no "node name" in the configuration.  Could just pull hostname of the machine though...

razuz commented:
I guess pulling hostname would be rather dirty ... 
@jbremer maybe it's time to add node_id / node_name setting to cuckoo.conf ?  or make this kind of var to reporting.conf  under [elasticsearch] ?

jbremer commented:
An optional option for the hostname would probably be good to have.
Then if it's not present, it shouldn't be included in the ES report either.
I think putting it in `cuckoo.conf` would be the cleanest - thoughts?

razuz commented:
+1 to cuckoo.conf - it will very likely come handy in other situations aswell

razuz commented:
FYI, 
not sure if you've also bumped into such issue, but I did it with .jar -s. running ES 2.3.3 and Cuckoo latest master with this PR

Cuckoo side:
----
WARNING:elasticsearch:POST /cuckoo-2016-06-22/cuckoo?op_type=create [status:400 request:0.005s]
WARNING:lib.cuckoo.core.plugins:The reporting module "ElasticSearch" returned the following error: Failed to save results in ElasticSearch for task #203: TransportError(400, u'mapper_parsing_exception', u'failed to parse [process.calls.arguments.type]')

ES side :
-----
[2016-06-22 03:51:14,808][DEBUG][action.index             ] u0001\u0000)(Ljava/util/concurrent/locks/Condition;)I\u0001\u0000)(Ljava/util/concurrent/locks/Condition;)Z\u0001\u0000#()Ljava/util/concurrent/locks/Lock;\u0001\u00006(Ljava/util/concurrent/locks/ReentrantReadWriteLock;)V\u0001\u0000>()Ljava/util/concurrent/locks/ReentrantReadWriteLock$ReadLock;\u0001\u0000?()Ljava/util/concurrent/locks/ReentrantReadWriteLock$WriteLock;\u0001\u0000-(Ljava/lang/String;)Ljava/lang/StringBuilder;\u0001\u0000,()Ljava/util/Collection<Ljava/lang/Thread;>;\u0001\u0000_(Ljava/util/concurrent/locks/AbstractQueuedSynchronizer$ConditionObject;)Ljava/util/Collection;\u0001\u0000>(Ljava/util/concurrent/locks/Condition;)Ljava/util/Collection;\u0001\u0000R(Ljava/util/concurrent/locks/Condition;)Ljava/util/Collection<Ljava/lang/Thread;>;\f\u0000=\u0000T\f\u0000@\u0000U\f\u0000C\u0000V\f\u0000\u0018\u0000\u0005\f\u0000\u001b\u0000\u0005\f\u0000\u001f\u0000\u0005\f\u0000 \u0000\u0005\f\u0000$\u0000\u0005\f\u0000\r\u0000\u0007\f\u0000&\u0000\b\f\u0000)\u0000\b\f\u0000+\u0000\b\f\u0000\u0017\u0000\t\f\u0000?\u0000\t\f\u0000\r\u0000\u000b\f\u0000A\u0000\u0006\f\u0000\r\u0000\n\f\u0000\u0016\u0000W\f\u0000\u001a\u0000X\f\u0000*\u0000Y\f\u0000\u0019\u0000Z\f\u0000\u001d\u0000Z\f\u0000!\u0000Z\f\u0000\"\u0000[\f\u0000'\u0000\\\f\u0000\r\u0000`\f\u0000<\u0000a\f\u0000B\u0000b\f\u0000\u0016\u0000c\f\u0000#\u0000e\t\u0000N\u0000h\t\u0000N\u0000i\t\u0000N\u0000j\n\u0000G\u0000x\n\u0000H\u0000p\n\u0000I\u0000p\n\u0000I\u0000w\n\u0000J\u0000p\n\u0000J\u0000w\n\u0000J\u0000y\n\u0000J\u0000\n\u0000N\u0000v\n\u0000N\u0000\n\u0000N\u0000\n\u0000O\u0000p\n\u0000P\u0000p\n\u0000Q\u0000\n\u0000R\u0000k\n\u0000R\u0000l\n\u0000R\u0000m\n\u0000R\u0000n\n\u0000R\u0000o\n\u0000R\u0000q\n\u0000R\u0000r\n\u0000R\u0000s\n\u0000R\u0000t\n\u0000R\u0000u\n\u0000R\u0000z\n\u0000R\u0000{\n\u0000R\u0000|\n\u0000R\u0000}\n\u0000R\u0000~\n\u0000R\u0000\n\u0000R\u0000\n\u0000R\u0000\n\u0000S\u0000\u0001\u0000\u000fConditionObject\u0001\u0000\bFairSync\u0001\u0000\u000bNonfairSync\u0001\u0000\bReadLock\u0001\u0000\u0004Sync\u0001\u0000\tWriteLock\u0000!\u0000N\u0000I\u0000\u0002\u0000M\u0000F\u0000\u0004\u0000\u001a\u0000>\u0000\u0011\u0000\u0001\u0000\u000f\u0000\u0000\u0000\u0002\u0000D\u0000\u0012\u0000=\u0000T\u0000\u0000\u0000\u0012\u0000C\u0000V\u0000\u0000\u0000\u0010\u0000@\u0000U\u0000\u0000\u0000\u0017\u0000\u0001\u0000\r\u0000\u0007\u0000\u0001\u0000\u000e\u0000\u0000\u0000\u0012\u0000\u0002\u0000\u0001\u0000\u0000\u0000\u0006*\u0003·\u0000±\u0000\u0000\u0000\u0000\u0000\u0001\u0000\r\u0000\u000b\u0000\u0001\u0000\u000e\u0000\u0000\u0000i\u0000\u0004\u0000\u0002\u0000\u0000\u00006*·\u0000*\u001b\u0000\r»\u0000OY·\u0000§\u0000\n»\u0000PY·\u0000µ\u0000*»\u0000QY*·\u0000µ\u0000*»\u0000SY*·\u0000©µ\u0000±\u0000\u0000\u0000\u0001\u0000\u0013\u0000\u0000\u0000!\u0000\u0002ÿ\u0000\u0013\u0000\u0002\u0007\u0000N\u0001\u0000\u0001\u0007\u0000Nÿ\u0000\u0006\u0000\u0002\u0007\u0000N\u0001\u0000\u0002\u0007\u0000N\u0007\u0000R\u0000\u0001\u0000B\u0000b\u0000\u0001\u0000\u000e\u0000\u0000\u0000\u0011\u0000\u0001\u0000\u0001\u0000\u0000\u0000\u0005*´\u0000°\u0000\u0000\u0000\u0000\u0000\u0001\u0000<\u0000a\u0000\u0001\u0000\u000e\u0000\u0000\u0000\u0011\u0000\u0001\u0000\u0001\u0000\u0000\u0000\u0005*´\u0000°\u0000\u0000\u0000\u0000\u0000\u0011\u0000(\u0000\b\u0000\u0001\u0000\u000e\u0000\u0000\u0000\u0014\u0000\u0001\u0000\u0001\u0000\u0000\u0000\b*´\u0000Á\u0000O¬\u0000\u0000\u0000\u0000\u0000\u0004\u0000\u001a\u0000X\u0000\u0001\u0000\u000e\u0000\u0000\u0000\u0014\u0000\u0001\u0000\u0001\u0000\u0000\u0000\b*´\u0000¶\u0000¡°\u0000\u0000\u0000\u0000\u0000\u0001\u0000 \u0000\u0005\u0000\u0001\u0000\u000e\u0000\u0000\u0000\u0014\u0000\u0001\u0000\u0001\u0000\u0000\u0000\b*´\u0000¶\u0000¬\u0000\u0000\u0000\u0000\u0000\u0001\u0000+\u0000\b\u0000\u0001\u0000\u000e\u0000\u0000\u0000\u0014\u0000\u0001\u0000\u0001\u0000\u0000\u0000\b*´\u0000¶\u0000¬\u0000\u0000\u0000\u0000\u0000\u0001\u0000,\u0000\b\u0000\u0001\u0000\u000e\u0000\u0000\u0000\u0014\u0000\u0001\u0000\u0001\u0000\u0000\u0000\b*´\u0000¶\u0000¬\u0000\u0000\u0000\u0000\u0000\u0001\u0000$\u0000\u0005\u0000\u0001\u0000\u000e\u0000\u0000\u0000\u0014\u0000\u0001\u0000\u0001\u0000\u0000\u0000\b*´\u0000¶\u0000¬\u0000\u0000\u0000\u0000\u0000\u0001\u0000\u001f\u0000\u0005\u0000\u0001\u0000\u000e\u0000\u0000\u0000\u0014\u0000\u0001\u0000\u0001\u0000\u0000\u0000\b*´\u0000¶\u0000¬\u0000\u0000\u0000\u0000\u0000\u0004\u0000\u001e\u0000Z\u0000\u0002\u0000\u000e\u0000\u0000\u0000\u0014\u0000\u0001\u0000\u0001\u0000\u0000\u0000\b*´\u0000¶\u0000£°\u0000\u0000\u0000\u0000\u0000\u0012\u0000\u0000\u0000\u0002\u0000d\u0000\u0004\u0000\u001c\u0000Z\u0000\u0002\u0000\u000e\u0000\u0000\u0000\u0014\u0000\u0001\u0000\u0001\u0000\u0000\u0000\b*´\u0000¶\u0000¥°\u0000\u0000\u0000\u0000\u0000\u0012\u0000\u0000\u0000\u0002\u0000d\u0000\u0011\u0000&\u0000\b\u0000\u0001\u0000\u000e\u0000\u0000\u0000\u0014\u0000\u0001\u0000\u0001\u0000\u0000\u0000\b*´\u0000¶\u0000¬\u0000\u0000\u0000\u0000\u0000\u0011\u0000%\u0000Y\u0000\u0001\u0000\u000e\u0000\u0000\u0000\u0015\u0000\u0002\u0000\u0002\u0000\u0000\u0000\t*´\u0000+¶\u0000¢¬\u0000\u0000\u0000\u0000\u0000\u0011\u0000\u001b\u0000\u0005\u0000\u0001\u0000\u000e\u0000\u0000\u0000\u0014\u0000\u0001\u0000\u0001\u0000\u0000\u0000\b*´\u0000¶\u0000¬\u0000\u0000\u0000\u0000\u0000\u0004\u0000\u001d\u0000Z\u0000\u0002\u0000\u000e\u0000\u0000\u0000\u0014\u0000\u0001\u0000\u0001\u0000\u0000\u0000\b*´\u0000¶\u0000¤°\u0000\u0000\u0000\u0000\u0000\u0012\u0000\u0000\u0000\u0002\u0000d\u0000\u0001\u0000'\u0000^\u0000\u0001\u0000\u000e\u0000\u0000\u0000?\u0000\u0003\u0000\u0002\u0000\u0000\u0000)+Ç\u0000\u000b»\u0000HY·\u0000¿+Á\u0000L\u0000\r»\u0000GY\u0012\u0004·\u0000¿*´\u0000+À\u0000L¶\u0000§¬\u0000\u0000\u0000\u0001\u0000\u0013\u0000\u0000\u0000\u0004\u0000\u0002\f\u0010\u0000\u0001\u0000\"\u0000]\u0000\u0001\u0000\u000e\u0000\u0000\u0000?\u0000\u0003\u0000\u0002\u0000\u0000\u0000)+Ç\u0000\u000b»\u0000HY·\u0000¿+Á\u0000L\u0000\r»\u0000GY\u0012\u0004·\u0000¿*´\u0000+À\u0000L¶\u0000¦¬\u0000\u0000\u0000\u0001\u0000\u0013\u0000\u0000\u0000\u0004\u0000\u0002\f\u0010\u0000\u0004\u0000#\u0000f\u0000\u0002\u0000\u000e\u0000\u0000\u0000?\u0000\u0003\u0000\u0002\u0000\u0000\u0000)+Ç\u0000\u000b»\u0000HY·\u0000¿+Á\u0000L\u0000\r»\u0000GY\u0012\u0004·\u0000¿*´\u0000+À\u0000L¶\u0000¨°\u0000\u0000\u0000\u0001\u0000\u0013\u0000\u0000\u0000\u0004\u0000\u0002\f\u0010\u0000\u0012\u0000\u0000\u0000\u0002\u0000g\u0000\u0001\u0000A\u0000\u0006\u0000\u0001\u0000\u000e\u0000\u0000\u0000G\u0000\u0002\u0000\u0004\u0000\u0000\u0000;*´\u0000¶\u0000<\u001b¸\u0000=\u001b¸\u0000 >»\u0000JY·\u0000*·\u0000¶\u0000\u0012\u0002¶\u0000\u001c¶\u0000\u0012\u0001¶\u0000\u001d¶\u0000\u0012\u0003¶\u0000¶\u0000°\u0000\u0000\u0000\u0000\u0010A\u0000B\u0000_\u0000\u0001\u0000\u000e\u0000\u0000\u0000\u0011\u0000\u0001\u0000\u0001\u0000\u0000\u0000\u0005*¶\u0000°\u0000\u0000\u0000\u0000\u0010A\u0000<\u0000_\u0000\u0001\u0000\u000e\u0000\u0000\u0000\u0011\u0000\u0001\u0000\u0001\u0000\u0000\u0000\u0005*¶\u0000°\u0000", "length": 4157, "offset": 0}, "time": "2016-06-22T06:40:45.484000", "tid": 2576, "flags": {}}, {"category": "file", "status": 1, "stacktrace": [], "api": "SetFilePointerEx", "return_value": 1, "arguments": {"file_handle": "0x000000b4", "move_method": 0, "offset": 50887584}, "time": "2016-06-22T06:40:45.484000", "tid": 2576, "flags": {}}, {"category": "file", "status": 1, "stacktrace": [], "api": "NtReadFile", "return_value": 0, "arguments": {"file_handle": "0x000000b4", "buffer": "PK\u0001\u0002\n\u0000\n\u0000\u0000\u0000\u0000\u0000\u0017\u0014Û>\u0001T,©\u0000\u0000\u0000©\u0000\u0000\u0000.\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000Q=|\u0001java/util/concurrent/locks/ReadWriteLock.classPK\u0001\u0002\n\u0000\n\u0000\u0000\u0000\u0000\u0000\u0015\u0014Û>és5Æ\u0003\u0000\u0000Æ\u0003\u0000\u00007\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000\u0000F>|\u0001java/util/concurrent/l", "length": 160, "offset": 0}, "time": "2016-06-22T06:40:45.484000", "tid": 2576, "flags": {}}, {"category": "file", "status": 1, "stacktrace": [], "api": "SetFilePointerEx", "return_value": 1, "arguments": {"file_handle": "0x000000b4", "move_method": 0, "offset": 24919377}, "time": "2016-06-22T06:40:45.484000", "tid": 2576, "flags": {}}, {"category": "file", "status": 1, "stacktrace": [], "api": "NtReadFile", "return_value": 0, "arguments": {"file_handle": "0x000000b4", "buffer": "PK\u0003\u0004\n\u0000\u0000\u0000\u0000\u0000\u0017\u0014Û>\u0001T,©\u0000\u0000\u0000©\u0000\u0000\u0000.\u0000\u0000\u0000", "length": 30, "offset": 0}, "time": "2016-06-22T06:40:45.484000", "tid": 2576, "flags": {}}, {"category": "file", "status": 1, "stacktrace": [], "api": "SetFilePointerEx", "return_value": 1, "arguments": {"file_handle": "0x000000b4", "move_method": 0, "offset": 24919453}, "time": "2016-06-22T06:40:45.484000", "tid": 2576, "flags": {}}, {"category": "file", "status": 1, "stacktrace": [], "api": "NtReadFile", "return_value": 0, "arguments": {"file_handle": "0x000000b4", "buffer": "Êþº¾\u0000\u0000\u00003\u0000\b\u0001\u0000\u0010java/lang/Object\u0001\u0000(java/util/concurrent/locks/ReadWriteLock\u0001\u0000\breadLock\u0001\u0000\twriteLock\u0007\u0000\u0001\u0007\u0000\u0002\u0001\u0000#()Ljava/util/concurrent/locks/Lock;\u0006\u0001\u0000\u0006\u0000\u0005\u0000\u0000\u0000\u0000\u0000\u0002\u0004\u0001\u0000\u0003\u0000\u0007\u0000\u0000\u0004\u0001\u0000\u0004\u0000\u0007\u0000\u0000\u0000\u0000", "length": 169, "offset": 0}, "time": "2016-06-22T06:40:45.484000", "tid": 2576, "flags": {}}, {"category": "file", "status": 1, "stacktrace": [], "api": "SetFilePointerEx", "return_value": 1, "arguments": {"file_handle": "0x000000b4", "move_method": 0, "offset": 50887887}, "time": "2016-06-22T06:40:45.484000", "tid": 2576, "flags": {}}]}, "report_time": 1466556619, "task_id": 203}]}] on [[cuckoo-2016-06-22][0]]
MapperParsingException[failed to parse [process.calls.arguments.type]]; nested: NumberFormatException[For input string: "AWT_ICON"];

swackhamer commented:
Do you have the behavioral option turned on?  https://github.com/swackhamer/cuckoo/blob/ebbc6284e6446f0bf2b7e461e4a8ac02159fda52/conf/reporting.conf#L32

Could you post the output of this command?

`curl` -XGET 'http://localhost:9200/cuckoo-*/_mappings'

razuz commented:
yep, I have calls = yes 

I took  only todays mapping :
`{"cuckoo-2016-06-22":{"mappings":{"cuckoo":{"dynamic_templates":[{"notanalyzed":{"mapping":{"index":"not_analyzed","doc_values":"True","type":"string"},"match":"*","match_mapping_type":"string"}}],"properties":{"process":{"properties":{"calls":{"properties":{"api":{"type":"string","index":"not_analyzed"},"arguments":{"properties":{"access":{"type":"string","index":"not_analyzed"},"af":{"type":"long"},"allocation_type":{"type":"long"},"application_name":{"type":"string","index":"not_analyzed"},"base_address":{"type":"string","index":"not_analyzed"},"base_handle":{"type":"string","index":"not_analyzed"},"basename":{"type":"string","index":"not_analyzed"},"buffer":{"type":"string","index":"not_analyzed"},"callback_function":{"type":"string","index":"not_analyzed"},"child_after_hwnd":{"type":"string","index":"not_analyzed"},"class":{"type":"string","index":"not_analyzed"},"class_context":{"type":"long"},"class_max_length":{"type":"long"},"class_name":{"type":"string","index":"not_analyzed"},"clsid":{"type":"string","index":"not_analyzed"},"command_line":{"type":"string","index":"not_analyzed"},"commit_size":{"type":"long"},"computer_name":{"type":"string","index":"not_analyzed"},"console_handle":{"type":"string","index":"not_analyzed"},"container":{"type":"string","index":"not_analyzed"},"control_code":{"type":"long"},"create_disposition":{"type":"long"},"create_options":{"type":"long"},"creation_flags":{"type":"long"},"crypto_handle":{"type":"string","index":"not_analyzed"},"current_directory":{"type":"string","index":"not_analyzed"},"database_name":{"type":"string","index":"not_analyzed"},"desired_access":{"type":"string","index":"not_analyzed"},"device_handle":{"type":"string","index":"not_analyzed"},"directory_handle":{"type":"string","index":"not_analyzed"},"dirpath":{"type":"string","index":"not_analyzed"},"dirpath_r":{"type":"string","index":"not_analyzed"},"disposition":{"type":"long"},"file_attributes":{"type":"long"},"file_handle":{"type":"string","index":"not_analyzed"},"file_size":{"type":"long"},"filepath":{"type":"string","index":"not_analyzed"},"filepath_r":{"type":"string","index":"not_analyzed"},"flags":{"type":"long"},"folder":{"type":"long"},"free_type":{"type":"long"},"function_address":{"type":"string","index":"not_analyzed"},"function_name":{"type":"string","index":"not_analyzed"},"handle":{"type":"string","index":"not_analyzed"},"handle_attributes":{"type":"long"},"hook_identifier":{"type":"long"},"hostname":{"type":"string","index":"not_analyzed"},"id":{"type":"long"},"iid":{"type":"string","index":"not_analyzed"},"index":{"type":"long"},"info_level":{"type":"long"},"information_class":{"type":"long"},"inherit_handles":{"type":"long"},"initial_owner":{"type":"long"},"input_buffer":{"type":"string","index":"not_analyzed"},"internet_handle":{"type":"string","index":"not_analyzed"},"key_handle":{"type":"string","index":"not_analyzed"},"key_name":{"type":"string","index":"not_analyzed"},"length":{"type":"long"},"library":{"type":"string","index":"not_analyzed"},"machine_name":{"type":"string","index":"not_analyzed"},"milliseconds":{"type":"long"},"mode":{"type":"long"},"module":{"type":"string","index":"not_analyzed"},"module_address":{"type":"string","index":"not_analyzed"},"module_handle":{"type":"string","index":"not_analyzed"},"module_name":{"type":"string","index":"not_analyzed"},"move_method":{"type":"long"},"mutant_handle":{"type":"string","index":"not_analyzed"},"mutant_name":{"type":"string","index":"not_analyzed"},"object_handle":{"type":"string","index":"not_analyzed"},"offset":{"type":"long"},"open_options":{"type":"long"},"option":{"type":"long"},"options":{"type":"long"},"ordinal":{"type":"long"},"output_buffer":{"type":"string","index":"not_analyzed"},"owner_handle":{"type":"string","index":"not_analyzed"},"parameter":{"type":"string","index":"not_analyzed"},"parent_hwnd":{"type":"string","index":"not_analyzed"},"privilege_name":{"type":"string","index":"not_analyzed"},"process_handle":{"type":"string","index":"not_analyzed"},"process_identifier":{"type":"long"},"processor_count":{"type":"long"},"protection":{"type":"long"},"protocol":{"type":"long"},"provider":{"type":"string","index":"not_analyzed"},"provider_type":{"type":"long"},"query":{"type":"string","index":"not_analyzed"},"query_language":{"type":"string","index":"not_analyzed"},"reg_type":{"type":"long"},"region_size":{"type":"long"},"regkey":{"type":"string","index":"not_analyzed"},"regkey_r":{"type":"string","index":"not_analyzed"},"resource_name":{"type":"string","index":"not_analyzed"},"section_handle":{"type":"string","index":"not_analyzed"},"section_name":{"type":"string","index":"not_analyzed"},"section_offset":{"type":"long"},"service_handle":{"type":"string","index":"not_analyzed"},"service_manager_handle":{"type":"string","index":"not_analyzed"},"service_name":{"type":"string","index":"not_analyzed"},"share_access":{"type":"long"},"shortpath":{"type":"string","index":"not_analyzed"},"size":{"type":"long"},"skipped":{"type":"long"},"socket":{"type":"long"},"source_handle":{"type":"string","index":"not_analyzed"},"source_process_handle":{"type":"string","index":"not_analyzed"},"source_process_identifier":{"type":"long"},"stack_size":{"type":"long"},"stack_zero_bits":{"type":"long"},"status_code":{"type":"string","index":"not_analyzed"},"status_info":{"type":"long"},"string":{"type":"string","index":"not_analyzed"},"subkey_count":{"type":"long"},"subkey_max_length":{"type":"long"},"suspend_count":{"type":"long"},"suspended":{"type":"long"},"system_name":{"type":"string","index":"not_analyzed"},"target_handle":{"type":"string","index":"not_analyzed"},"target_process_handle":{"type":"string","index":"not_analyzed"},"target_process_identifier":{"type":"long"},"thread_handle":{"type":"string","index":"not_analyzed"},"thread_identifier":{"type":"long"},"thread_name":{"type":"string","index":"not_analyzed"},"token_handle":{"type":"string","index":"not_analyzed"},"track":{"type":"long"},"type":{"type":"long"},"username":{"type":"string","index":"not_analyzed"},"uuid":{"type":"string","index":"not_analyzed"},"value":{"type":"string","index":"not_analyzed"},"value_count":{"type":"long"},"value_max_length":{"type":"long"},"value_name_max_length":{"type":"long"},"view_size":{"type":"long"},"volume_mount_point":{"type":"string","index":"not_analyzed"},"volume_name":{"type":"string","index":"not_analyzed"},"volume_path_name":{"type":"string","index":"not_analyzed"},"wVersionRequested":{"type":"long"},"win32_protect":{"type":"long"},"window_name":{"type":"string","index":"not_analyzed"}}},"category":{"type":"string","index":"not_analyzed"},"flags":{"properties":{"allocation_type":{"type":"string","index":"not_analyzed"},"clsid":{"type":"string","index":"not_analyzed"},"control_code":{"type":"string","index":"not_analyzed"},"create_disposition":{"type":"string","index":"not_analyzed"},"create_options":{"type":"string","index":"not_analyzed"},"creation_flags":{"type":"string","index":"not_analyzed"},"desired_access":{"type":"string","index":"not_analyzed"},"file_attributes":{"type":"string","index":"not_analyzed"},"folder":{"type":"string","index":"not_analyzed"},"hook_identifier":{"type":"string","index":"not_analyzed"},"iid":{"type":"string","index":"not_analyzed"},"index":{"type":"string","index":"not_analyzed"},"information_class":{"type":"string","index":"not_analyzed"},"mode":{"type":"string","index":"not_analyzed"},"open_options":{"type":"string","index":"not_analyzed"},"option":{"type":"string","index":"not_analyzed"},"protection":{"type":"string","index":"not_analyzed"},"reg_type":{"type":"string","index":"not_analyzed"},"share_access":{"type":"string","index":"not_analyzed"},"status_info":{"type":"string","index":"not_analyzed"},"win32_protect":{"type":"string","index":"not_analyzed"}}},"last_error":{"type":"long"},"nt_status":{"type":"long"},"return_value":{"type":"long"},"status":{"type":"long"},"tid":{"type":"long"},"time":{"type":"date","format":"strict_date_optional_time||epoch_millis"}}},"page":{"type":"long"},"pid":{"type":"long"}}},"report_time":{"type":"date","format":"epoch_second"},"summary":{"properties":{"command_line":{"type":"string","index":"not_analyzed"},"connects_ip":{"type":"string","index":"not_analyzed"},"directory_created":{"type":"string","index":"not_analyzed"},"directory_enumerated":{"type":"string","index":"not_analyzed"},"dll_loaded":{"type":"string","index":"not_analyzed"},"downloads_file":{"type":"string","index":"not_analyzed"},"file_created":{"type":"string","index":"not_analyzed"},"file_deleted":{"type":"string","index":"not_analyzed"},"file_exists":{"type":"string","index":"not_analyzed"},"file_failed":{"type":"string","index":"not_analyzed"},"file_opened":{"type":"string","index":"not_analyzed"},"file_read":{"type":"string","index":"not_analyzed"},"file_recreated":{"type":"string","index":"not_analyzed"},"file_written":{"type":"string","index":"not_analyzed"},"guid":{"type":"string","index":"not_analyzed"},"mutex":{"type":"string","index":"not_analyzed"},"regkey_deleted":{"type":"string","index":"not_analyzed"},"regkey_opened":{"type":"string","index":"not_analyzed"},"regkey_read":{"type":"string","index":"not_analyzed"},"regkey_written":{"type":"string","index":"not_analyzed"},"resolves_host":{"type":"string","index":"not_analyzed"},"wmi_query":{"type":"string","index":"not_analyzed"}}},"target":{"properties":{"category":{"type":"string","index":"not_analyzed"},"file":{"properties":{"crc32":{"type":"string","index":"not_analyzed"},"md5":{"type":"string","index":"not_analyzed"},"name":{"type":"string","index":"not_analyzed"},"path":{"type":"string","index":"not_analyzed"},"sha1":{"type":"string","index":"not_analyzed"},"sha256":{"type":"string","index":"not_analyzed"},"sha512":{"type":"string","index":"not_analyzed"},"size":{"type":"long"},"type":{"type":"string","index":"not_analyzed"},"yara":{"properties":{"meta":{"properties":{"author":{"type":"string","index":"not_analyzed"},"description":{"type":"string","index":"not_analyzed"}}},"name":{"type":"string","index":"not_analyzed"},"strings":{"type":"string","index":"not_analyzed"}}}}}}},"task_id":{"type":"long"},"virustotal":{"properties":{"md5":{"type":"string","index":"not_analyzed"},"normalized":{"type":"string","index":"not_analyzed"},"permalink":{"type":"string","index":"not_analyzed"},"positives":{"type":"long"},"resource":{"type":"string","index":"not_analyzed"},"response_code":{"type":"long"},"scan_date":{"type":"string","index":"not_analyzed"},"scan_id":{"type":"string","index":"not_analyzed"},"scans":{"properties":{"ALYac":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"AVG":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"AVware":{"properties":{"detected":{"type":"boolean"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Ad-Aware":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"AegisLab":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Alibaba":{"properties":{"detected":{"type":"boolean"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Antiy-AVL":{"properties":{"detected":{"type":"boolean"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Arcabit":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Avast":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Avira":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Baidu":{"properties":{"detected":{"type":"boolean"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Baidu-International":{"properties":{"detected":{"type":"boolean"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"BitDefender":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Bkav":{"properties":{"detected":{"type":"boolean"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"CAT-QuickHeal":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"CMC":{"properties":{"detected":{"type":"boolean"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"ClamAV":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Comodo":{"properties":{"detected":{"type":"boolean"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Cyren":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"DrWeb":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"ESET-NOD32":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Emsisoft":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"F-Prot":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"F-Secure":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Fortinet":{"properties":{"detected":{"type":"boolean"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"GData":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Ikarus":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Jiangmin":{"properties":{"detected":{"type":"boolean"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"K7AntiVirus":{"properties":{"detected":{"type":"boolean"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"K7GW":{"properties":{"detected":{"type":"boolean"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Kaspersky":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Kingsoft":{"properties":{"detected":{"type":"boolean"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Malwarebytes":{"properties":{"detected":{"type":"boolean"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"McAfee":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"McAfee-GW-Edition":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"MicroWorld-eScan":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Microsoft":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"NANO-Antivirus":{"properties":{"detected":{"type":"boolean"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Panda":{"properties":{"detected":{"type":"boolean"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Qihoo-360":{"properties":{"detected":{"type":"boolean"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"SUPERAntiSpyware":{"properties":{"detected":{"type":"boolean"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Sophos":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Symantec":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Tencent":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"TheHacker":{"properties":{"detected":{"type":"boolean"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"TotalDefense":{"properties":{"detected":{"type":"boolean"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"TrendMicro":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"TrendMicro-HouseCall":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"VBA32":{"properties":{"detected":{"type":"boolean"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"VIPRE":{"properties":{"detected":{"type":"boolean"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"ViRobot":{"properties":{"detected":{"type":"boolean"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Yandex":{"properties":{"detected":{"type":"boolean"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Zillya":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"Zoner":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"ahnlab":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}},"nProtect":{"properties":{"detected":{"type":"boolean"},"normalized":{"type":"string","index":"not_analyzed"},"result":{"type":"string","index":"not_analyzed"},"update":{"type":"string","index":"not_analyzed"},"version":{"type":"string","index":"not_analyzed"}}}}},"sha1":{"type":"string","index":"not_analyzed"},"sha256":{"type":"string","index":"not_analyzed"},"summary":{"properties":{"error":{"type":"string","index":"not_analyzed"},"permalink":{"type":"string","index":"not_analyzed"},"positives":{"type":"long"},"scan_date":{"type":"string","index":"not_analyzed"}}},"total":{"type":"long"},"verbose_msg":{"type":"string","index":"not_analyzed"}}}}}}}}`

swackhamer commented:
Yep that is exactly what is happening.  Your "process.calls.arguments.type" is set to a long.

Do you have any dynamic templates that could be affecting the mappings?  Such as from another Cuckoo sandbox fork or other custom code?

If not then I am not sure how exactly to handle this.  For some reason process.calls.arguments.type is being interrupted as a number for you.  We need to verify that the python dict submitted on this line:https://github.com/swackhamer/cuckoo/blob/ebbc6284e6446f0bf2b7e461e4a8ac02159fda52/modules/reporting/elasticsearch.py#L99  is the right python literal and will never change from string to int.

Or I could add the full mapping to the json template and then if a document is parsed with a int the first time it comes in but is for a string field ES will just convert that int to a string.  The only issue is if the schema changes in later builds the template will need to be updated.  I was trying to stay away from that.

razuz commented:
I had dynamic template on cuckoo, but flushed it for now. This had a different issue.
I will play around with that more in couple days and debug it out.

swackhamer commented:
If you wanted to run this command to delete the index template that was applied with this pull request:

    `curl -XDELETE 'https://es.mumblederp.com/_template/cuckoo_template'`

and then swap out template.json with this file: https://gist.github.com/swackhamer/5cf7a19649e410478eed147b3ae19a7b

That will explicitly specify the type of all the mappings.  I am a little bit hesitant to include a 2755 line mapping file.  I think this may be a issue more into Cuckoo internals as well where a dynamically typed field is sometimes a number and sometimes a string.

razuz commented:
I presume sticking with dynamic makes more sense as this is just one sample and theres probably more :)

razuz commented:
how about using template like that :
`{
    "order": 0,
    "template": "cuckoo-*",
    "settings": {
        "index": {
            "number_of_shards": "1",
            "codec": "best_compression",
            "number_of_replicas": "0"
        }
    },
    "mappings": {
        "cuckoo": {
            "dynamic_templates": [
                {
                    "notanalyzed": {
                        "mapping": {
                            "index": "not_analyzed",
                            "type": "string",
                            "doc_values": "True"
                        },
                        "match_mapping_type": "string",
                        "match": "*"
                    }
                },
                {
                    "long_to_string": {
                        "mapping": {
                            "index": "not_analyzed",
                            "type": "string",
                            "doc_values": "True"
                        },
                        "match_mapping_type": "long",
                        "match": "*"
                    }
                }
            ],
            "properties": {
                "report_time": {
                    "format": "epoch_second",
                    "type": "date"
                },
                "task_id": {
                    "type": "long"
                }
            }
        }
    },
    "aliases": {}
}
`

this takes :
* long fields and maps them as strings (For calls I guess this could be the best way and yeah it's rather hack for longs, but solves issue I experience) 
* use 0 replica by default - this can be changed by user depending on the ES cluster setup

"match_mapping_type" : "string|long" - seems like this doesn't work as expected at least on ES 2.3.3 I have for sandbox

one other note here - this PR breaks the search as analysis ID is now date not ID

swackhamer commented:
Made some major changes:

- Integrated Razu's changes and set replica to 0 in the index template so users without a distributed elasticsearch cluster will not have unallocated shards
- In code converted all call arguments to strings so we should not have any mapping issue as all the data is being inserted into elasticsearch as a indexable string
- Added in the option to index yearly, monthly, or daily.  Yearly templates are set by default
- Added in bulk submission support which should speed up indexing of the big behavioral alerts
- Changed the ways calls are being indexed -- no more pagination and every call has its own document.  This allows Kibana to easily visualize on the call data now too.
- Fixed the cuckoo web interface search feature, should be better then ever!  Also it does not search the full ES cluster anymore just the Cuckoo index

jbremer commented:
> Changed the ways calls are being indexed -- no more pagination and every call has its own document. This allows Kibana to easily visualize on the call data now too.

Does that scale?

swackhamer commented:
Elasticsearch handles very small documents very well and with the change to using the bulk API to bulk submit 10k call documents at a time this change has greatly increased the scalability.  There is a small overhead with adding some document meta data (report_time, report_id, pid, etc...) versus indexing a completely new report but I think this is the most "elasticsearch" way to store this data.  It is less "normalized".

Indexing calls into elasticsearch is a pretty time intensive prospect.  I have been profiling the code and on my machine with calls turned on it is around 120 seconds for ~800k calls (office documents with a lot of calls).  I have been able to shrink that using multiprocessing to submit into Elasticsearch to ~70 seconds.  Unfortunately at the end of the day you are sending a ton of raw data to elasticsearch.

Indexing calls is very interesting and I think it will lead to some very cool results, but for most Cuckoo installs it may be best to keep it off as it is currently in the config right now.

Here is a interesting Kibana visualization on one report that produced 840k calls:
![cuckoo-kibana](https://cloud.githubusercontent.com/assets/619624/16361157/1ae51d7c-3b43-11e6-83b9-1c12ea97d17e.PNG)

jbremer commented:
Keeping the mass indexing in mind I think this could be nice. However, I'll wait with merging for some feedback from @razuz. And, as discussed, `multiprocessing`-based processing will not be a thing for now.

razuz commented:
there's a typo in https://github.com/cuckoosandbox/cuckoo/pull/968/files#diff-2b8b7dc997960f8b3760bd47db2e4ec7R43
It should be index_time_pattern rather than index_pattern ... 

other than that it looks good to me 👍 in case anything pops up I'll create a patch

Once the PR is merged I'll PR the node_id var also to cuckoo.conf and adopt it with ES

swackhamer commented:
Change made.

jbremer commented:
Thanks a lot @swackhamer @razuz!

jbremer commented:
Btw, @razuz @swackhamer, shouldn't each API call document contain the `call index`? So it can be linked directly to the correct behavioral page later on when searching through ElasticSearch?

swackhamer commented:
Here is an example pagination ES query script that can lookup api calls by page number.  There is no unique primary key for the specific call, ES does autogen a ID for it but of course that is not stored anywhere in Cuckoo.  I don't think that is going to be a problem for anything though.

https://gist.github.com/swackhamer/eb3a03f7cd2580370fc4c89b64b2f4be
